package com.bmt.kaleidoscope_nether.datagen;

import com.bmt.kaleidoscope_nether.KaleidoscopeNether;
import com.bmt.kaleidoscope_nether.init.KNBlocks;
import com.github.ysbbbbbb.kaleidoscopecookery.block.food.FoodBiteBlock;
import com.github.ysbbbbbb.kaleidoscopecookery.init.registry.FoodBiteRegistry;

import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.data.PackOutput;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.CaveVinesBlock;
import net.minecraft.world.level.block.CropBlock;
import net.neoforged.neoforge.client.model.generators.BlockStateProvider;
import net.neoforged.neoforge.client.model.generators.ConfiguredModel;
import net.neoforged.neoforge.client.model.generators.ModelFile;
import net.neoforged.neoforge.client.model.generators.VariantBlockStateBuilder;
import net.neoforged.neoforge.common.data.ExistingFileHelper;
import net.neoforged.neoforge.registries.DeferredHolder;

public class ModBlockStateProvider extends BlockStateProvider {
    public ModBlockStateProvider(PackOutput output, ExistingFileHelper exFileHelper) {
        super(output, KaleidoscopeNether.MOD_ID, exFileHelper);
    }

    @Override
    protected void registerStatesAndModels() {
        cropBlock(KNBlocks.POISONOUS_FRUIT);
        cropBlock(KNBlocks.SOUL_PEPPER);

        FoodBiteRegistry.FOOD_DATA_MAP.forEach((resourceLocation, foodData) -> {
            if (resourceLocation.getNamespace().equals(KaleidoscopeNether.MOD_ID)) {
                Block block = BuiltInRegistries.BLOCK.get(resourceLocation);
                addFoodBiteBlock(block, resourceLocation);
            }
        });


        caveVines(KNBlocks.TWISTING_CAVE_VINES);
        caveVines(KNBlocks.TWISTING_CAVE_VINES_PLANT);

        caveVines(KNBlocks.WEEPING_CAVE_VINES);
        caveVines(KNBlocks.WEEPING_CAVE_VINES_PLANT);
    }


    protected void caveVines(DeferredHolder<Block, ? extends Block> holder) {
        Block block = holder.get();
        VariantBlockStateBuilder builder = getVariantBuilder(block);
        builder.addModels(builder.partialState().with(CaveVinesBlock.BERRIES, true), ConfiguredModel.builder().modelFile(new ModelFile.UncheckedModelFile(modLoc("block/%s_berries".formatted(holder.getId().getPath())))).build());
        builder.addModels(builder.partialState().with(CaveVinesBlock.BERRIES, false), ConfiguredModel.builder().modelFile(new ModelFile.UncheckedModelFile(modLoc("block/%s".formatted(holder.getId().getPath())))).build());
    }

    protected void cropBlock(DeferredHolder<Block, ? extends Block> holder) {
        Block block = holder.get();
        VariantBlockStateBuilder builder = getVariantBuilder(block);
        builder.forAllStates(blockState -> {
            int age = blockState.getValue(CropBlock.AGE);
            ResourceLocation file = modLoc("block/%s/stage%d".formatted(holder.getId().getPath(), age));
            return ConfiguredModel.builder().modelFile(new ModelFile.UncheckedModelFile(file)).build();
        });
    }

    public void addFoodBiteBlock(Block block, ResourceLocation id) {
        this.horizontalBlock(block, (blockState) -> {
            if (block instanceof FoodBiteBlock foodBiteBlock) {
                int bites = blockState.getValue(foodBiteBlock.getBites());
                ResourceLocation model = KaleidoscopeNether.fromNamespaceAndPath(id.getNamespace(), "block/food/%s/%s_%d".formatted(id.getPath(), id.getPath(), bites));
                return new ModelFile.UncheckedModelFile(model);
            } else {
                throw new IllegalArgumentException("Block must be an instance of FoodBiteBlock");
            }
        });
    }
}
